File: src/shared/ipc-contracts/story-preparation.ts
/**
 * Wire contract for the story-preparation flow. Mirror of
 * `src-tauri/src/ipc/dto/story_preparation.rs` (state DTO + acceptance) and
 * `src-tauri/src/ipc/events.rs` (the three `job:*` event payloads).
 *
 * The preparation state is COMPOSED by Rust; the frontend only PRESENTS it. The
 * `preflight` / `preparing` variants describe the in-flight phases the hook
 * derives from `job:progress` events; `read_preparation_state` itself only ever
 * returns `idle` / `prepared` / `retryable`. Cross-stack contract tests keep the
 * shapes symmetric.
 */

import { isValidationBlocker, type ValidationBlocker } from "./story-validation";

/** Closed set of functional preparation-failure causes (AC3). */
export type PreparationCause =
  | "preflightNotPassing"
  | "artifactMissing"
  | "artifactCorrupt"
  | "deviceChanged"
  | "interrupted";

/** The selected local story identified in the preparation state. */
export interface PreparationStory {
  /** Local story id (canonical lowercase UUID). */
  id: string;
  /** The local `stories.title`. */
  title: string;
}

export type PreparationStateDto =
  | { kind: "idle" }
  | { kind: "preflight"; deviceIdentifier: string; story: PreparationStory }
  | {
      kind: "preparing";
      deviceIdentifier: string;
      story: PreparationStory;
      progress: number | null;
    }
  | {
      kind: "prepared";
      deviceIdentifier: string;
      story: PreparationStory;
      targetCohort: string;
      /** Whether the prepared descriptor carries a device-format pack (an
       *  imported story). `false` for a native story with no pack — the send
       *  gate disables `Envoyer` before any write attempt. */
      transferable: boolean;
    }
  | {
      kind: "retryable";
      story: PreparationStory;
      cause: PreparationCause;
      message: string;
      userAction: string;
      blockers: ValidationBlocker[];
    };

/** Acceptance returned by `start_prepare_story`. */
export interface StartPreparationAcceptedDto {
  jobId: string;
  storyId: string;
}

/** Phase carried by a `job:progress` event (the in-flight phases). Shared by
 *  the preparation flow (`preflight` / `prepare`) and the transfer flow
 *  (`transfer`, on the same job channel). `verify` stays out of scope. */
export type JobPhase = "preflight" | "prepare" | "transfer";

export interface JobProgressEvent {
  jobId: string;
  jobType: string;
  targetStoryId: string;
  phase: JobPhase;
  progress: number | null;
  sequence: number;
  message: string | null;
}

export interface JobCompletedEvent {
  jobId: string;
  jobType: string;
  targetStoryId: string;
  sequence: number;
}

export interface JobFailedEvent {
  jobId: string;
  jobType: string;
  targetStoryId: string;
  sequence: number;
  errorCode: string;
  errorMessage: string;
  userAction: string;
  /** Transfer-only (AC2): whether the device was already mutated when the write
   *  failed — `"failed"` (device left untouched → `échec récupérable`) vs
   *  `"incomplete"` (the write had started → `transfert incomplet`). Absent for
   *  preparation jobs, which have no device-mutation notion. */
  completeness?: "failed" | "incomplete";
  /** Transfer-only (AC3): the structured failure cause (camelCase) so the UI keeps
   *  "cause + issue + next action" in context, not only the message. Absent for
   *  preparation and the non-classifiable defensive terminal. */
  cause?: string;
}

const CAUSES: ReadonlySet<string> = new Set([
  "preflightNotPassing",
  "artifactMissing",
  "artifactCorrupt",
  "deviceChanged",
  "interrupted",
]);

// The generic job channel accepts every live phase Rustory emits: `preflight` /
// `prepare` (preparation) and `transfer` (the write flow). `verify` is reserved
// and stays rejected until its flow exists.
const PHASES: ReadonlySet<string> = new Set(["preflight", "prepare", "transfer"]);

const DEVICE_IDENTIFIER_PATTERN = /^[0-9a-f]{32}$/;
const STORY_ID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

// A Map (not a plain object): `.get(kind)` returns `undefined` for an unknown OR
// inherited key, so `hasOnlyAllowedKeys` reports drift instead of throwing.
const ALLOWED_KEYS: ReadonlyMap<string, ReadonlySet<string>> = new Map([
  ["idle", new Set(["kind"])],
  ["preflight", new Set(["kind", "deviceIdentifier", "story"])],
  ["preparing", new Set(["kind", "deviceIdentifier", "story", "progress"])],
  [
    "prepared",
    new Set(["kind", "deviceIdentifier", "story", "targetCohort", "transferable"]),
  ],
  [
    "retryable",
    new Set(["kind", "story", "cause", "message", "userAction", "blockers"]),
  ],
]);

const ALLOWED_STORY_KEYS: ReadonlySet<string> = new Set(["id", "title"]);

const ALLOWED_PROGRESS_KEYS: ReadonlySet<string> = new Set([
  "jobId",
  "jobType",
  "targetStoryId",
  "phase",
  "progress",
  "sequence",
  "message",
]);
const ALLOWED_COMPLETED_KEYS: ReadonlySet<string> = new Set([
  "jobId",
  "jobType",
  "targetStoryId",
  "sequence",
]);
const ALLOWED_FAILED_KEYS: ReadonlySet<string> = new Set([
  "jobId",
  "jobType",
  "targetStoryId",
  "sequence",
  "errorCode",
  "errorMessage",
  "userAction",
  "completeness",
  "cause",
]);

const COMPLETENESS: ReadonlySet<string> = new Set(["failed", "incomplete"]);

function hasOnlyAllowedKeys(
  value: Record<string, unknown>,
  allowed: ReadonlySet<string> | undefined,
): boolean {
  if (!allowed) return false;
  for (const k of Object.keys(value)) {
    if (!allowed.has(k)) return false;
  }
  return true;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}

function isPreparationStory(value: unknown): value is PreparationStory {
  if (typeof value !== "object" || value === null) return false;
  const s = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(s, ALLOWED_STORY_KEYS)) return false;
  if (typeof s.id !== "string" || !STORY_ID_PATTERN.test(s.id)) return false;
  if (typeof s.title !== "string" || s.title.length === 0) return false;
  return true;
}

function isDeviceIdentifier(value: unknown): value is string {
  return typeof value === "string" && DEVICE_IDENTIFIER_PATTERN.test(value);
}

function isProgress(value: unknown): value is number | null {
  return (
    value === null ||
    (typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 1)
  );
}

function isSequence(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= 0;
}

/**
 * Runtime guard for `PreparationStateDto`. Rejects every drift: unknown `kind`,
 * missing/extra fields, wrong types, malformed `deviceIdentifier` / `story.id`,
 * unrecognized `cause`, empty `message` / `userAction`, malformed blockers.
 */
export function isPreparationStateDto(
  value: unknown,
): value is PreparationStateDto {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (typeof c.kind !== "string") return false;
  if (!hasOnlyAllowedKeys(c, ALLOWED_KEYS.get(c.kind))) return false;
  switch (c.kind) {
    case "idle":
      return true;
    case "preflight":
      return isDeviceIdentifier(c.deviceIdentifier) && isPreparationStory(c.story);
    case "preparing":
      return (
        isDeviceIdentifier(c.deviceIdentifier) &&
        isPreparationStory(c.story) &&
        isProgress(c.progress)
      );
    case "prepared":
      return (
        isDeviceIdentifier(c.deviceIdentifier) &&
        isPreparationStory(c.story) &&
        isNonEmptyString(c.targetCohort) &&
        typeof c.transferable === "boolean"
      );
    case "retryable":
      if (!isPreparationStory(c.story)) return false;
      if (typeof c.cause !== "string" || !CAUSES.has(c.cause)) return false;
      if (!isNonEmptyString(c.message)) return false;
      if (!isNonEmptyString(c.userAction)) return false;
      if (!Array.isArray(c.blockers)) return false;
      // Reuse the 3.x blocker guard verbatim — never two wordings for one cause.
      return c.blockers.every(isValidationBlocker);
    default:
      return false;
  }
}

/** Runtime guard for the acceptance returned by `start_prepare_story`. Both ids
 *  are canonical lowercase UUIDs (the `jobId` is a generated UUID, the `storyId`
 *  the selected story) — an empty / malformed id would have the hook subscribe to
 *  an impossible correlation, so it is rejected as drift like the other guards. */
export function isStartPreparationAcceptedDto(
  value: unknown,
): value is StartPreparationAcceptedDto {
  if (typeof value !== "object" || value === null) return false;
  const a = value as Record<string, unknown>;
  if (Object.keys(a).length !== 2) return false;
  return (
    typeof a.jobId === "string" &&
    STORY_ID_PATTERN.test(a.jobId) &&
    typeof a.storyId === "string" &&
    STORY_ID_PATTERN.test(a.storyId)
  );
}

function isJobBaseShape(c: Record<string, unknown>): boolean {
  return (
    isNonEmptyString(c.jobId) &&
    isNonEmptyString(c.jobType) &&
    typeof c.targetStoryId === "string" &&
    isSequence(c.sequence)
  );
}

export function isJobProgressEvent(value: unknown): value is JobProgressEvent {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, ALLOWED_PROGRESS_KEYS)) return false;
  if (!isJobBaseShape(c)) return false;
  if (typeof c.phase !== "string" || !PHASES.has(c.phase)) return false;
  if (!isProgress(c.progress)) return false;
  return c.message === null || typeof c.message === "string";
}

export function isJobCompletedEvent(value: unknown): value is JobCompletedEvent {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, ALLOWED_COMPLETED_KEYS)) return false;
  return isJobBaseShape(c);
}

export function isJobFailedEvent(value: unknown): value is JobFailedEvent {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, ALLOWED_FAILED_KEYS)) return false;
  if (!isJobBaseShape(c)) return false;
  // `completeness` is optional (transfer-only); when present it must be a known
  // variant.
  if (c.completeness !== undefined && !COMPLETENESS.has(c.completeness as string)) {
    return false;
  }
  // `cause` is optional (transfer-only); when present it must be a non-empty
  // string (the closed transfer cause set lives in the transfer contract).
  if (c.cause !== undefined && !isNonEmptyString(c.cause)) {
    return false;
  }
  return (
    isNonEmptyString(c.errorCode) &&
    isNonEmptyString(c.errorMessage) &&
    isNonEmptyString(c.userAction)
  );
}

File: src-tauri/tests/contracts/story_preparation.rs
//! Wire-shape contract for the story-preparation flow: the exact JSON of every
//! `PreparationStateDto` variant, the acceptance DTO, and the three `job:*`
//! event payloads. The frontend mirror (`src/shared/ipc-contracts/
//! story-preparation.ts`) must stay byte-compatible with these shapes.

use rustory_lib::application::transfer::PreparationStateView;
use rustory_lib::domain::story::{Axis, CanonicalBlocker, CanonicalCause, Severity};
use rustory_lib::domain::transfer::PreparationFailureCause;
use rustory_lib::ipc::dto::{
    PreparationStateDto, PreparationStoryDto, StartPreparationAcceptedDto,
};
use rustory_lib::ipc::events::{
    JobCompletedEvent, JobFailedEvent, JobProgressEvent, JOB_TYPE_PREPARE_STORY,
    PREPARATION_FAILED_CODE,
};
use serde_json::json;

const DEVICE: &str = "0123456789abcdef0123456789abcdef";
const STORY: &str = "0197a5d0-0000-7000-8000-000000000000";

fn story() -> PreparationStoryDto {
    PreparationStoryDto {
        id: STORY.into(),
        title: "Mon histoire".into(),
    }
}

#[test]
fn idle_is_a_single_kind_key() {
    let v = serde_json::to_value(PreparationStateDto::Idle).expect("ser");
    assert_eq!(v, json!({ "kind": "idle" }));
}

#[test]
fn preflight_and_preparing_wire_shapes() {
    let v = serde_json::to_value(PreparationStateDto::Preflight {
        device_identifier: DEVICE.into(),
        story: story(),
    })
    .expect("ser");
    assert_eq!(v["kind"], "preflight");
    assert_eq!(v["deviceIdentifier"], DEVICE);
    assert_eq!(v["story"]["id"], STORY);

    let v = serde_json::to_value(PreparationStateDto::Preparing {
        device_identifier: DEVICE.into(),
        story: story(),
        progress: None,
    })
    .expect("ser");
    assert_eq!(v["kind"], "preparing");
    assert_eq!(v["progress"], json!(null));
}

#[test]
fn prepared_wire_shape_is_camel_case() {
    let dto = PreparationStateDto::from_view(PreparationStateView::Prepared {
        device_identifier: DEVICE.into(),
        story_id: STORY.into(),
        story_title: "Mon histoire".into(),
        target_cohort: "v3".into(),
        transferable: false,
    });
    let v = serde_json::to_value(&dto).expect("ser");
    assert_eq!(v["kind"], "prepared");
    assert_eq!(v["deviceIdentifier"], DEVICE);
    assert_eq!(v["story"]["title"], "Mon histoire");
    assert_eq!(v["targetCohort"], "v3");
    assert_eq!(v["transferable"], false);
    assert!(v.get("target_cohort").is_none(), "no snake_case leak");
}

#[test]
fn retryable_wire_shape_carries_non_empty_action_and_blockers() {
    let dto = PreparationStateDto::from_view(PreparationStateView::Retryable {
        story_id: STORY.into(),
        story_title: "Mon histoire".into(),
        cause: PreparationFailureCause::PreflightNotPassing,
        blockers: vec![CanonicalBlocker {
            axis: Axis::Structure,
            cause: CanonicalCause::ChecksumMismatch,
            severity: Severity::Blocking,
        }],
    });
    let v = serde_json::to_value(&dto).expect("ser");
    assert_eq!(v["kind"], "retryable");
    assert_eq!(v["cause"], "preflightNotPassing");
    assert!(!v["message"].as_str().expect("message").is_empty());
    assert!(!v["userAction"].as_str().expect("userAction").is_empty());
    assert_eq!(v["blockers"][0]["axis"], "structure");
    assert_eq!(v["blockers"][0]["cause"], "checksumMismatch");
    assert!(!v["blockers"][0]["userAction"]
        .as_str()
        .expect("blocker userAction")
        .is_empty());
}

#[test]
fn every_cause_serializes_to_a_distinct_camel_case_discriminant() {
    let mut seen = Vec::new();
    for cause in [
        PreparationFailureCause::PreflightNotPassing,
        PreparationFailureCause::ArtifactMissing,
        PreparationFailureCause::ArtifactCorrupt,
        PreparationFailureCause::DeviceChanged,
        PreparationFailureCause::Interrupted,
    ] {
        let dto = PreparationStateDto::from_view(PreparationStateView::Retryable {
            story_id: STORY.into(),
            story_title: "T".into(),
            cause,
            blockers: vec![],
        });
        let v = serde_json::to_value(&dto).expect("ser");
        let tag = v["cause"].as_str().expect("cause").to_string();
        assert!(!tag.is_empty());
        seen.push(tag);
    }
    let mut unique = seen.clone();
    unique.sort();
    unique.dedup();
    assert_eq!(
        unique.len(),
        seen.len(),
        "causes must be distinct on the wire"
    );
}

#[test]
fn acceptance_dto_wire_shape() {
    let v = serde_json::to_value(StartPreparationAcceptedDto {
        job_id: "job-1".into(),
        story_id: STORY.into(),
    })
    .expect("ser");
    assert_eq!(v, json!({ "jobId": "job-1", "storyId": STORY }));
}

#[test]
fn job_progress_event_wire_shape() {
    let v = serde_json::to_value(JobProgressEvent {
        job_id: "j1".into(),
        job_type: JOB_TYPE_PREPARE_STORY.into(),
        target_story_id: STORY.into(),
        phase: "preflight".into(),
        progress: None,
        sequence: 1,
        message: None,
    })
    .expect("ser");
    assert_eq!(
        v,
        json!({
            "jobId": "j1",
            "jobType": "prepare_story",
            "targetStoryId": STORY,
            "phase": "preflight",
            "progress": null,
            "sequence": 1,
            "message": null,
        })
    );
}

#[test]
fn job_completed_event_wire_shape() {
    let v = serde_json::to_value(JobCompletedEvent {
        job_id: "j1".into(),
        job_type: JOB_TYPE_PREPARE_STORY.into(),
        target_story_id: STORY.into(),
        sequence: 3,
    })
    .expect("ser");
    assert_eq!(v["jobId"], "j1");
    assert_eq!(v["jobType"], "prepare_story");
    assert_eq!(v["targetStoryId"], STORY);
    assert_eq!(v["sequence"], 3);
}

#[test]
fn job_failed_event_carries_non_empty_message_and_action() {
    let v = serde_json::to_value(JobFailedEvent {
        job_id: "j1".into(),
        job_type: JOB_TYPE_PREPARE_STORY.into(),
        target_story_id: STORY.into(),
        sequence: 2,
        error_code: PREPARATION_FAILED_CODE.into(),
        error_message: "Préparation interrompue.".into(),
        user_action: "Relance la préparation.".into(),
        completeness: None,
        cause: None,
    })
    .expect("ser");
    assert_eq!(v["errorCode"], "PREPARATION_FAILED");
    assert!(
        v.get("completeness").is_none(),
        "preparation failures carry no completeness"
    );
    assert!(!v["errorMessage"].as_str().expect("message").is_empty());
    assert!(!v["userAction"].as_str().expect("userAction").is_empty());
    assert!(v.get("error_code").is_none(), "no snake_case leak");
}
